### ISSUE TITLE
Floating Table of contents on docs pages

### ISSUE THREAD

**Issue by papkos** [django/djangoproject.com#239]
It would be really nice if the TOC on the right side of every documentation page could "float" down if I scroll the main page. It's quite time-wasting to have to scroll all the way up if I quickly want to jump to another section on the same page.

I know that the TOC is often longer than the screen's height, but I belive this could be worked around. Maybe a screen-height sized container containing the scrollable TOC? 

I really miss this feature (and missed it on the old site too)!

**Comment by timgraham** [comment]
Thanks for the suggestion, but I think this will add more complexity for questionable benefit as Ola mentioned.

**Comment by jacklinke** [comment]
Looks like #203 was resolved via #1370, and I have just proposed a solution for #494, so moving up **and** down a docs page quickly may soon be easy and this closed Issue may remain unnecessary, but I'm sharing this chunk of code in case anyone deems this Issue worth opening back up. I'm no front-end expert, but I was curious how difficult this would be, and got this working relatively quickly.

On larger screens, it makes the sidebar float (resizing and adjusting the top as needed if the header or footer are visible) and makes the sidebar content independently scrollable.

It should work on most browsers, but again, not a front-end expert and have not tested outside of the browsers I have installed locally.

```html
<style>
  @media screen and (min-width: 768px) {
    [role="complementary"] {
      position: fixed;
      top: 160px;
      right: 20px;
      max-height: calc(100vh - 20px);
      overflow-y: scroll;
      transition: top 0.3s ease; /* Smooth the transition for top value */
    }

    /* Hide scrollbar but still allow scrolling */
    [role="complementary"]::-webkit-scrollbar {
      width: 0;
    }
    [role="complementary"] {
      scrollbar-width: none; /* Firefox */
    }
  }
</style>

<script>
    let timeout;
    const complementaryElement = document.querySelector('[role="complementary"]');

    function calculateHeaderHeight() {
      // Calculate the height of the header and copy banner
      const topHeight = document.getElementById('top').offsetHeight;
      const copyBannerHeight = document.querySelector('#generic > div.copy-banner').offsetHeight;
      return topHeight + copyBannerHeight;
    }

    function calculateFooterHeight() {
      // Calculate the height of the footer
      return document.querySelector('#generic > div:nth-child(6)').offsetHeight;
    }

    function updateComplementaryPosition(complementaryElement) {
      // Update the sidebar position based on the scroll position
      const headerHeight = calculateHeaderHeight();
      const footerHeight = calculateFooterHeight();
      
      const scrollY = window.scrollY;
      const body = document.body;
      const html = document.documentElement;
      const totalPageHeight = Math.max( body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight );
      const scrollablePageHeight = totalPageHeight - window.innerHeight;

      if (window.scrollY < headerHeight) {  // We're near the top of the page
        const newTop = Math.max(headerHeight - scrollY, 0);
        complementaryElement.style.top = `${newTop}px`;
      } else if (window.scrollY > scrollablePageHeight - footerHeight) {  // We're near the bottom of the page
        const newMaxHeight = Math.max(window.innerHeight - (scrollY - (scrollablePageHeight - footerHeight)) - 20, 0)
        complementaryElement.style['max-height'] = `${newMaxHeight}px`;
      } else {  // We're somewhere in the middle of the page
        const newMaxHeight = Math.max(window.innerHeight - 20, 0)
        complementaryElement.style['max-height'] = `${newMaxHeight}px`;
        complementaryElement.style.top = 0 + "px";
      }
    }

    function initialComplementaryPosition(complementaryElement) {
      // Set the initial position of the sidebar
      const headerHeight = calculateHeaderHeight();
      const scrollY = window.scrollY;
      const newTop = Math.max(headerHeight - scrollY, 0);
      complementaryElement.style.top = `${newTop}px`;
    }

    // Add event listener for scroll event to update the sidebar position as we scroll
    document.addEventListener("scroll", () => {
      // Add a timeout to debounce a bit when scrolling a lot
      if (timeout) clearTimeout(timeout);
      timeout = setTimeout(() => {
        if (complementaryElement) {
          updateComplementaryPosition(complementaryElement);
        }
      }, 50);
    });

    // Set the initial position of the sidebar
    if (complementaryElement) {
      initialComplementaryPosition(complementaryElement);
    }
</script>
```
